File: ios/Runner/SensorApiImplementation.swift
//
//  SensorApiImplementation.swift
//  Runner
//
//  Created by Aleksander Kapera on 10/13/24.
//

import Foundation
import CoreBluetooth

class SensorApiImplementation: NSObject, SensorApi, ObservableObject, IBluetoothEventObserver, IBwt901bleRecordObserver, CBCentralManagerDelegate {
    
    // Get bluetooth manager
    var bluetoothManager: WitBluetoothManager = WitBluetoothManager.instance
    private var centralManager: CBCentralManager!
    private var stateCallback: ((Result<Bool, Error>) -> Void)?
    
    fileprivate var flutterApi: FlutterApi
    
    // Bluetooth 5.0 sensor object
    @Published
    var deviceList:[Bwt901ble] = [Bwt901ble]()
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        self.flutterApi = FlutterApi(binaryMessenger: binaryMessenger)
    }
    
    func initialize(completion: @escaping (Result<Bool, Error>) -> Void) {
        centralManager = CBCentralManager(delegate: self, queue: nil)
        self.stateCallback = completion
    }
    
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
            case .poweredOn:
                stateCallback?(Result.success(true))
                break
            default:
                stateCallback?(Result.success(false))
                break
        }
    }
    
    func startDiscovery() throws {
        print("Start scanning for surrounding bluetooth devices")
        // Remove all devices, here all devices are turned off and removed from the list
        removeAllDevice()
        
        // Registering a Bluetooth event observer
        self.bluetoothManager.registerEventObserver(observer: self)
        
        // Turn on bluetooth scanning
        self.bluetoothManager.startScan()
    }
    
    // MARK: This method is called if a Bluetooth Low Energy sensor is found
    func onFoundBle(bluetoothBLE: BluetoothBLE?) {
        if isNotFound(bluetoothBLE) {
            print("\(String(describing: bluetoothBLE?.peripheral.name)) found a bluetooth device")
            self.deviceList.append(Bwt901ble(bluetoothBLE: bluetoothBLE))
            openDevice(bwt901ble: Bwt901ble(bluetoothBLE: bluetoothBLE))
        }
    }
    
    // Judging that the device has not been found
    func isNotFound(_ bluetoothBLE: BluetoothBLE?) -> Bool{
        for device in deviceList {
            if device.mac == bluetoothBLE?.mac {
                return false
            }
        }
        return true
    }
    
    // MARK: You will be notified here when the connection is successful
    func onConnected(bluetoothBLE: BluetoothBLE?) {
        print("\(String(describing: bluetoothBLE?.peripheral.name)) connected")
        flutterApi.onSensorConnected(isConnected: true)
    }
    
    // MARK: Notifies you here when the connection fails
    func onConnectionFailed(bluetoothBLE: BluetoothBLE?) {
        print("\(String(describing: bluetoothBLE?.peripheral.name)) failed to connect")
        flutterApi.onSensorConnected(isConnected: false)
    }
    
    // MARK: You will be notified here when the connection is lost
    func onDisconnected(bluetoothBLE: BluetoothBLE?) {
        print("\(String(describing: bluetoothBLE?.peripheral.name)) disconnected")
        flutterApi.onSensorConnected(isConnected: false)
    }
    
    // MARK: Stop scanning for devices
    func stopScan(){
        self.bluetoothManager.removeEventObserver(observer: self)
        self.bluetoothManager.stopScan()
    }
    
    func openDevice(bwt901ble: Bwt901ble?){
        print("Turn on the device")
        
        do {
            try bwt901ble?.openDevice()
            
            // Monitor data
            bwt901ble?.registerListenKeyUpdateObserver(obj: self)
            
            /**
             * 0x01：0.1Hz
             * 0x02：0.5Hz
             * 0x03：1Hz
             * 0x04：2Hz
             * 0x05：5Hz
             * 0x06：10Hz（default）
             * 0x07：20Hz
             * 0x08：50Hz
             * 0x09：100Hz
             * 0x0A：200Hz
             */
            try bwt901ble?.unlockReg()
            try bwt901ble?.writeRge([0xff ,0xaa, 0x03, 0x08, 0x00], 10)
            try bwt901ble?.saveReg()
        }
        catch{
            print("Failed to open device")
        }
    }
    
    // MARK: Remove all devices
    func removeAllDevice(){
        for item in deviceList {
            closeDevice(bwt901ble: item)
        }
        deviceList.removeAll()
    }
    
    // MARK: Turn off the device
    func closeDevice(bwt901ble: Bwt901ble?){
        print("Turn off the device")
        bwt901ble?.closeDevice()
        flutterApi.onSensorConnected(isConnected: false)
    }
    
    
    // MARK: You will be notified here when data from the sensor needs to be recorded
    func onRecord(_ bwt901ble: Bwt901ble) {
        // You can get sensor data here
        let deviceData = getDeviceData(bwt901ble)
        flutterApi.onSensorDataRecorded(sensorData: deviceData)
    }
    
    // MARK: Get the data of the device and concatenate it into a string
    func getDeviceData(_ device:Bwt901ble) -> SensorDataModel {
        return SensorDataModel(name: device.name ?? "",
                          acceleration: ThreeAxisMeasurementModel(
                            x: Double(device.getDeviceData(WitSensorKey.AccX) ?? ""),
                            y: Double(device.getDeviceData(WitSensorKey.AccY) ?? ""),
                            z: Double(device.getDeviceData(WitSensorKey.AccZ) ?? "")),
                          angularVelocity: ThreeAxisMeasurementModel(
                            x: Double(device.getDeviceData(WitSensorKey.GyroX) ?? ""),
                            y: Double(device.getDeviceData(WitSensorKey.GyroY) ?? ""),
                            z: Double(device.getDeviceData(WitSensorKey.GyroZ) ?? "")),
                          magneticField: ThreeAxisMeasurementModel(
                            x: Double(device.getDeviceData(WitSensorKey.MagX) ?? ""),
                            y: Double(device.getDeviceData(WitSensorKey.MagY) ?? ""),
                            z: Double(device.getDeviceData(WitSensorKey.MagZ) ?? "")),
                          angle: ThreeAxisMeasurementModel(
                            x: Double(device.getDeviceData(WitSensorKey.AngleX) ?? ""),
                            y: Double(device.getDeviceData(WitSensorKey.AngleY) ?? ""),
                            z: Double(device.getDeviceData(WitSensorKey.AngleZ) ?? "")))
    }
}

private class FlutterApi {
    var flutterAPI: SensorCallbackApi
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        flutterAPI = SensorCallbackApi(binaryMessenger: binaryMessenger)
    }
    
    func onSensorDataRecorded(sensorData: SensorDataModel) {
        flutterAPI.onSensorDataRecorded(sensorData: sensorData) {_ in
        }
    }
    
    func onSensorConnected(isConnected: Bool){
        flutterAPI.onSensorConnected(isConnected: isConnected) {_ in
        }
    }
}

File: android/app/src/main/kotlin/app/getforma/formaApp/SensorApiImplementation.kt
package app.getforma.formaApp

import SensorApi
import SensorDataModel
import SensorCallbackpi
import ThreeAxisMeasurementModel
import androidx.app.Activity
import androidx.content.Context
import com.wit.witsdk.modular.sensor.device.exceptions.OpenDeviceException
import com.wit.witsdk.modular.sensor.example.ble5.Bwt901ble
import com.wit.witsdk.modular.sensor.example.ble5.interfaces.IBwt901bleRecordObserver
import com.wit.witsdk.modular.sensor.modular.connector.modular.bluetooth.BluetoothBLE
import com.wit.witsdk.modular.sensor.modular.connector.modular.bluetooth.BluetoothSPP
import com.wit.witsdk.modular.sensor.modular.connector.modular.bluetooth.WitBluetoothManager
import com.wit.witsdk.modular.sensor.modular.connector.modular.bluetooth.exceptions.BluetoothBLEException
import com.wit.witsdk.modular.sensor.modular.connector.modular.bluetooth.interfaces.IBluetoothFoundObserver
import com.wit.witsdk.modular.sensor.modular.processor.constant.WitSensorKey
import io.flutter.plugin.common.BinaryMessenger


class SensorApiImplementation private constructor(
    private val context: Context,
    messenger: BinaryMessenger,
    activity: Activity,
) : SensorApi,
    IBluetoothFoundObserver, IBwt901bleRecordObserver {

    /**
     * Device List
     */
    private val bwt901bleList: MutableList<Bwt901ble> = mutableListOf()

    private val flutterApi = FlutterApi(messenger, activity)

    companion object {

        @Volatile
        private var instance: SensorApi? = null

        fun getInstance(context: Context, messenger: BinaryMessenger, activity: Activity) =
            instance ?: synchronized(this) {
                instance ?: SensorApiImplementation(context, messenger, activity).also {
                    instance = it
                }
            }
    }

    override fun initialize() {
        WitBluetoothManager.initInstance(context)
    }

    override fun startDiscovery() {
        // Turn off all device
        bwt901bleList.forEach {
            it.removeRecordObserver(this)
            it.close()
        }

        // Erase all devices
        bwt901bleList.clear()

        // Start searching for bluetooth
        try {
            // get bluetooth manager
            val bluetoothManager = WitBluetoothManager.getInstance()

            // Monitor communication signals
            bluetoothManager.registerObserver(this)

            // Specify the Bluetooth name to search for
//            WitBluetoothManager.DeviceNameFilter = ["WT"]

            // start search
            bluetoothManager.startDiscovery()
        } catch (e: BluetoothBLEException) {
            e.printStackTrace()
        }
    }

    override fun onFoundBle(bluetoothBLE: BluetoothBLE?) {
        // Create a Bluetooth 5.0 sensor connection object
        val bwt901ble = Bwt901ble(bluetoothBLE)

        for (i in bwt901bleList.indices) {
            if (bwt901bleList[i].deviceName == bwt901ble.deviceName) {
                return
            }
        }

        // add to device list
        bwt901bleList.add(bwt901ble)

        // Registration data record
        bwt901ble.registerRecordObserver(this)

        // Turn on the device
        try {
            bwt901ble.open()
            setSensorFrequency(bwt901ble)
        } catch (e: OpenDeviceException) {
            // Failed to open device
            e.printStackTrace()
        }
    }

    override fun onFoundSPP(p0: BluetoothSPP?) {
        // Our sensor connects only via Bluetooth 5.0. This method is used for Bluetooth 2.0.
    }

    /**
     * This method will be called back when data needs to be recorded
     */
    override fun onRecord(bwt901ble: Bwt901ble?) {
        val deviceData = getDeviceData(bwt901ble) ?: return
        flutterApi.onSensorDataRecorded(sensorData = deviceData)
//        Log.d(
//            MainActivity::class.java.simpleName,
//            "device data = $deviceData"
//        )
    }

    private fun getDeviceData(bwt901ble: Bwt901ble?): SensorDataModel? {
        if (bwt901ble == null) {
            return null
        }

        return SensorDataModel(
            name = bwt901ble.deviceName,
            acceleration = ThreeAxisMeasurementModel(
                x = bwt901ble.getDeviceData(WitSensorKey.AccX)?.toDoubleOrNull(),
                y = bwt901ble.getDeviceData(WitSensorKey.AccY)?.toDoubleOrNull(),
                z = bwt901ble.getDeviceData(WitSensorKey.AccZ)?.toDoubleOrNull(),
            ),
            angularVelocity = ThreeAxisMeasurementModel(
                x = bwt901ble.getDeviceData(WitSensorKey.AsX)?.toDoubleOrNull(),
                y = bwt901ble.getDeviceData(WitSensorKey.AsY)?.toDoubleOrNull(),
                z = bwt901ble.getDeviceData(WitSensorKey.AsZ)?.toDoubleOrNull(),
            ),
            magneticField = ThreeAxisMeasurementModel(
                x = bwt901ble.getDeviceData(WitSensorKey.HX)?.toDoubleOrNull(),
                y = bwt901ble.getDeviceData(WitSensorKey.HY)?.toDoubleOrNull(),
                z = bwt901ble.getDeviceData(WitSensorKey.HZ)?.toDoubleOrNull(),
            ),
            angle = ThreeAxisMeasurementModel(
                x = bwt901ble.getDeviceData(WitSensorKey.AngleX)?.toDoubleOrNull(),
                y = bwt901ble.getDeviceData(WitSensorKey.AngleY)?.toDoubleOrNull(),
                z = bwt901ble.getDeviceData(WitSensorKey.AngleZ)?.toDoubleOrNull(),
            )
        )
    }

    private fun setSensorFrequency(bwt901ble: Bwt901ble) {
        /**
         * 0x01：0.1Hz
         * 0x02：0.5Hz
         * 0x03：1Hz
         * 0x04：2Hz
         * 0x05：5Hz
         * 0x06：10Hz（default）
         * 0x07：20Hz
         * 0x08：50Hz
         * 0x09：100Hz
         * 0x0A：200Hz
         */
        bwt901ble.setReturnRate(0x08)
    }
}

private class FlutterApi(messenger: BinaryMessenger, val activity: Activity) {

    val flutterApi: SensorCallbackpi = SensorCallbackApi(messenger)

    fun onSensorDataRecorded(sensorData: SensorData) {
        activity.runOnUiThread {
            flutterApi.onSensorDataRecorded(sensorData) { _ -> }
        }
    }
}
